### ISSUE TITLE
Show where a conflict is coming from

### ISSUE THREAD

**Issue by koterpillar** [jazzband/pip-tools#473]
When `pip-compile` is given conflicting requirements, it stops with a message about the conflict, but doesn't tell where is it coming from.

##### Environment Versions

1. OS Type: Linux
1. Python version: 2.7.13
1. pip version:  9.0.1
1. pip-tools version: 1.8.1

##### Steps to replicate

1. Have a requirements file with conflicting packages, for example:
```python
djangorestframework==3.2.*
drf-extensions
```
2. `pip-compile requirements.in`

##### Expected result

A message explaining the source of conflict is shown, for example:

```
djangorestframework==3.2.* conflicts with djangorestframework>=3.3.1 from drf-extensions
```

##### Actual result

A conflict message is shown, but doesn't mention which package(s) caused conflict - note that `drf-extensions` isn't mentioned. If there is a lot of packages in `requirements.in`, it can be tedious to find the source of conflict.

```
Could not find a version that matches djangorestframework==3.2.*,>=3.3.1
Tried: (snip)
```

**Comment by vphilippon** [comment]
I started to work on this.
A good part from the job comes from the fact that when merging requirements, we gave up on all information from where they came from, so it wasn't really possible to point out which packages requirements are in conflict.

But I think I found a way to keep track of the "parent".
You can see my progress on this branch: vphilippon/pip-tools@084c549aba9bc3d424296f7c36bd8208fa0aa809

**Comment by vphilippon** [comment]
Quick update: I haven't had much time to continue on this.

Also, even if we manage to show the tracking of the dependency, in the case where the are a lot of packages, its going to be a _mess_. Something like:
```
Could not find a version that matches djangorestframework==3.2.*,>=3.3.1
(from line X ->pkg_a vX.X.X->pkg_b vX.X.X->[pkg_c vX.X.X, pkg_d vX.X.X]->djangorestframework==3.2.*,>=3.3.1
```
and this wasn't a deep example. Not that it didn't even give the required version of `djangorestframework` by pkg_c vX.X.X and pkg_d vX.X.X seperately, as we can't (simply) do that, as we merged the requirements (a thing `pip` does not really support, which is why we where clearing the tracking information).

Doing a grep on a verbose output is the best way for tracking a conflict right now, and will still be the only way to get the "per-parent specifier berfore merge" even if I ever manage to get through this.

Unless someone brings in additional value for the tracking info I mentionned, it's likely I won't go through with this.

**Comment by jedie** [comment]
Think i have the same Problem...

I wanted to install Django v1.11.x with `django>=1.11,<1.12` and it seems that other packages doesn't support v1.11, yet... But i didn't find it via `grep on a verbose output`, because some many lines contains `django` ;)

Any idea?

**Comment by vphilippon** [comment]
@jedie There's a few things you can try out:
- Make a stricter regex (or re-grep the result) to get lines with something like `django.*<` to get the packages that put an upper bound on django. I really just eyeballed that regex example, you might need something better.
- Try with another `requirements.in`, removing some of the "suspects".  That's a guessing game.
I hope this helps you out a bit.

**Comment by asottile** [comment]
If you're curious, here's how a tool (that I've worked on) (which does a similar thing to pip-sync) [does this](https://github.com/Yelp/venv-update/blob/638f1c87f301ae01d6f61b40793e5281762c3492/pip_faster.py#L284-L331)

**Comment by georgek** [comment]
Fixed by #837
